Initial support for trmm and trtri

e10harvey opened the pull request:
Related to #589

This is the first out of two PRs. This PR introduces TPL specializations for `trmm` and `trtri`. The second PR will introduce fall-back specializations for both `trmm` and `trtri`.

This PR introduces the following when the `blas,cublas,magma` TPLs are enabled:
- `KokkosBlas::trmm` calls into either BLAS or CUBLAS.
- `KokkosBlas:trtri` calls into either BLAS or MAGMA(GPU).

The two unit tests, one for `trmm` and another for `trtri`, that have been added have been verified to pass for both CUDA/MAGMA and OpenMP using all supported scalar types, memory layouts, and memory spaces.

srajama1 commented:
@vqd8a @iyamazaki Can you please review this as well ? I am doing a quick review, but I would appreciate your input.

srajama1 commented on src/impl/tpls/KokkosBlasLapack_trtri_tpl_spec_avail.hpp:
```diff
@@ -0,0 +1,133 @@
+/*
+//@HEADER
+// ************************************************************************
```
I would call it just KokkosLAPACK.

mhoemmen commented on src/blas/KokkosBlas3_trmm.hpp:
```diff
@@ -0,0 +1,179 @@
+  int64_t B_n = B.extent(1);
+
+  // Return if degenerated matrices are provided
+  if((A_m == 0) || (A_n == 0) || (B_m == 0) || (B_n == 0))
```
You don't need the inner parentheses, and some compilers warn about them.  `||` has lower precedence than `==`.

mhoemmen commented on src/blas/KokkosBlasLapack_trtri.hpp:
```diff
@@ -0,0 +1,144 @@
+                 "AViewType must have rank 2.");
+
+  // Check validity of indicator argument
+  bool valid_uplo  = (uplo[0] == 'U' ) || (uplo[0] == 'u' )||
```
You don't need the inner parentheses here.

mhoemmen commented on src/blas/KokkosBlasLapack_trtri.hpp:
```diff
@@ -0,0 +1,144 @@
+  RViewInternalType R = RViewInternalType(&result);
+
+  KokkosBlas::Impl::TRTRI<RViewInternalType, AViewInternalType>::trtri (R, uplo, diag, A);
+  Kokkos::fence();
```
I get why you need a fence here, but it seems like users either don't care about the error code, or they want it left on device to avoid an unnecessary fence or (synchronous) copy back to host.  Your interface here forces that result back to host, even if users plan on ignoring the error code.

vqd8a commented on src/impl/tpls/KokkosBlas_tpl_spec.hpp:
```diff
@@ -18,11 +61,18 @@ struct CudaBlasSingleton {
 #ifdef KOKKOSKERNELS_ENABLE_TPL_MAGMA
 #include "magma_v2.h"
+#ifndef KOKKOSKERNELS_ENABLE_TPL_LAPACK
 #include "magma_lapack.h"
```
I think ```magma_v2.h``` would be enough. Please double check.

e10harvey replied:
TODO: see if `gesv` requires `magma_lapack.h`.

vqd8a commented on src/impl/tpls/KokkosBlas_Lapack_tpl.hpp:
```diff
@@ -0,0 +1,43 @@
+    HostBlas<double >::trtri(const int matrix_layout, const char uplo, const char diag,
+                                         int n, // TODO: n,lda lapack_int
+                                         const double *a, int lda) {
+      return LAPACKE_dtrtri(matrix_layout, uplo, diag, n, const_cast <double*> (a), lda);
```
Why would we need LAPACKE interfaces? Can we just use Fortran interfaces instead?

e10harvey replied:
Same as above. There are no fortran interfaces for `trtri`.

vqd8a replied:
@e10harvey I think there are. Please check this out https://software.intel.com/en-us/mkl-developer-reference-fortran-trtri. I used Fortran interfaces for ```gesv```

e10harvey commented on Makefile.kokkos-kernels:
```diff
@@ -27,6 +27,7 @@ KOKKOSKERNELS_INTERNAL_ENABLE_MAGMA := $(strip $(shell echo $(KOKKOSKERNELS_ENAB
 KOKKOSKERNELS_INTERNAL_ENABLE_CHOLMOD := $(strip $(shell echo $(KOKKOSKERNELS_ENABLE_TPLS) | grep "cholmod" | wc -l))
 KOKKOSKERNELS_INTERNAL_ENABLE_SUPERLU := $(strip $(shell echo $(KOKKOSKERNELS_ENABLE_TPLS) | grep "superlu" | wc -l))
 KOKKOSKERNELS_INTERNAL_ENABLE_LAPACKE := $(strip $(shell echo $(KOKKOSKERNELS_ENABLE_TPLS) | grep "lapacke" | wc -l))
```
> Should we remove lapacke and cblas ?

SPTRSV code is using LAPACKE and CBLAS defines. I would remove LAPACKE and CBLAS when we update the SPTRSV code to use `KokkosBlas::{trmm,trtri}` -- @srajama1: shall I update SPTRSV as part of this PR? I think @iyamazaki is working on SPTRSV right now so we may get ourselves into some conflicts if I update SPTRSV in this PR.

> @e10harvey And why would we need LAPACK? Shouldn't KOKKOSKERNELS_INTERNAL_ENABLE_BLAS be enough since you put both `trmm` and `trtri` in `KokkkosBlas` namespace?

@vqd8a: Where `trtri` is supported confused me. LAPACK supports `trtri` on the host while MAGMA supports `trtri` on the device. BLAS does not support `trtri`.  So for `trtri` TPL specializations we may use LAPACK or MAGMA. This is unfortunate as LAPACK and MAGMA are additional dependencies.

vqd8a replied:
I just afraid we have too many macros. Usually, MKL, OpenBLAS and MAGMA provide blas functions and lapack functions in a single library. Correct me if I am wrong.

e10harvey commented on src/impl/tpls/KokkosBlas3_trmm_tpl_spec_decl.hpp:
```diff
@@ -0,0 +1,301 @@
+//KOKKOSBLAS3_DTRMM_BLAS( Kokkos::LayoutRight, Kokkos::LayoutRight, Kokkos::HostSpace, true)
+KOKKOSBLAS3_DTRMM_BLAS( Kokkos::LayoutRight, Kokkos::LayoutRight, Kokkos::HostSpace, false)
+
+//KOKKOSBLAS3_STRMM_BLAS( Kokkos::LayoutLeft,  Kokkos::LayoutLeft,  Kokkos::HostSpace, true)
```
cut and paste --  I followed @vqd8a `trsm` code when adding support for `trmm`. These commented out MACROS are for the ETI/Fall-back specializations which have not been added yet.

e10harvey commented on src/blas/KokkosBlas3_trmm.hpp:
```diff
@@ -0,0 +1,179 @@
+                 "BViewType must have rank 2.");
+
+  // Check validity of indicator argument
+  bool valid_side  = (side[0] == 'L' ) || (side[0] == 'l' )||
```
:+1: I think enums will improve readability and reduce the number of conditionals here. Also, we're not trying to match the BLAS `trmm` interface exactly anyways.

e10harvey commented on src/blas/KokkosBlasLapack_trtri.hpp:
```diff
@@ -0,0 +1,144 @@
+/// \file KokkosBlasLapack_trtri.hpp
+
+#include "KokkosKernels_Macros.hpp"
+#include "KokkosBlasLapack_trtri_spec.hpp"
```
Agreed. I did want to make it clear that `trtri` depends on LAPACK but given that we now have support using MAGMA, renaming makes sense.

e10harvey commented on src/impl/tpls/KokkosBlas_Lapack_tpl.cpp:
```diff
@@ -0,0 +1,3 @@
+#include <Kokkos_Core.hpp>
+#include <KokkosKernels_config.h>
+#include<KokkosBlas_Lapack_tpl.hpp>
```
We need LAPACK for `trtri`. There are C symbol collisions between LAPACK and BLAS, which is the motivation for this separate cpp file.

vqd8a replied:
If we use Fortran interfaces, there should be no symbol collision.

e10harvey commented:
@srajama1, @mhoemmen, @vqd8a: thanks very much for your feedback!

>     2\. All those implementations in macros will one day cause y'all a lot of trouble.  It's getting close to write-only code.

These MACROS are definitely error prone. Tracking down a compiler error/warning points the developer to the line where the MACRO is invoked. But, the MACROS let the preprocessor copy-and-paste all the different SCALAR type, MEMORY space, etc permutations. The only alternative I can think of would entail creating many _slightly_ different copies of the templated classes. If we copied by hand, for trtri, we would have 12 copies of the TRTRI class instead of 2. With the fall-back implementation, we would have 24 copies of the TRTRI class. I think this should be revisited offline as a separate issue.

mhoemmen commented:
@e10harvey wrote:
> But, the MACROS let the preprocessor copy-and-paste all the different SCALAR type, MEMORY space, etc permutations.

That part is fine -- the issue is that you have an actual implementation in the macro, complete with backslashes for endlines.  The original idea was to use `extern template`-style ETI, which wouldn't require this "implementation in a macro" thing.  Even the C++98 - style ETI that Trilinos does, doesn't have this problem.

e10harvey commented on src/impl/tpls/KokkosBlasLapack_trtri_tpl_spec_decl.hpp:
```diff
@@ -0,0 +1,203 @@
+    \
+    bool A_is_layout_left = std::is_same<Kokkos::LayoutLeft,LAYOUTA>::value; \
+    \
+    magma_int_t AST = A_is_layout_left?A.stride(1):A.stride(0), LDA = (AST == 0) ? 1 : AST; \
```
The leading dimension of A argument handles this. This has been verified via the trtri unit test. I'm not sure how the layout will affect spatial locality but I'm just concerned with correctness for now.

e10harvey commented:
@vqd8a, @srajama1, @ndellingwood, @mhoemmen: I've addressed most of the PR feedback. Let me know if there's anything else that you'd like addressed. I will squash and merge once approved.

91326b0 passed the trmm and trtri unit tests for all types/layouts as well as the spot check:
```bash
Running on machine: white
Going to test compilers:  gcc/6.4.0 gcc/7.2.0 gcc/7.4.0 ibm/16.1.0 cuda/9.2.88 cuda/10.0.130
Testing compiler gcc/6.4.0
Testing compiler gcc/7.2.0
  Starting job gcc-6.4.0-OpenMP_Serial-release
  Starting job gcc-7.2.0-OpenMP-release
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-6.4.0-OpenMP_Serial-release
Testing compiler gcc/7.4.0
  Starting job gcc-7.2.0-OpenMP_Serial-release
  PASSED gcc-7.2.0-Serial-release
Testing compiler ibm/16.1.0
  Starting job gcc-7.4.0-OpenMP-release
  PASSED gcc-7.4.0-OpenMP-release
Testing compiler cuda/9.2.88
  Starting job ibm-16.1.0-Serial-release
  PASSED gcc-7.2.0-OpenMP_Serial-release
  PASSED ibm-16.1.0-Serial-release
Testing compiler cuda/10.0.130
  Starting job cuda-9.2.88-Cuda_OpenMP-release
  PASSED cuda-9.2.88-Cuda_OpenMP-release
  Starting job cuda-10.0.130-Cuda_Serial-release
  PASSED cuda-10.0.130-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.0.130-Cuda_Serial-release build_time=1086 run_time=7
cuda-9.2.88-Cuda_OpenMP-release build_time=1005 run_time=7
gcc-6.4.0-OpenMP_Serial-release build_time=517 run_time=5
gcc-7.2.0-OpenMP-release build_time=294 run_time=4
gcc-7.2.0-OpenMP_Serial-release build_time=564 run_time=9
gcc-7.2.0-Serial-release build_time=272 run_time=7
gcc-7.4.0-OpenMP-release build_time=304 run_time=4
ibm-16.1.0-Serial-release build_time=1022 run_time=4
#######################################################
FAILED TESTS
#######################################################
```

srajama1 commented:
@e10harvey : There seem to be some conflicts.

@vqd8a : Assuming the conflict will be addressed, is this ready to go in ? We could address our implementation in the next PR.

vqd8a commented:
@e10harvey Thanks for the fixing.
You switched to use ```KOKKOSKERNELS_ENABLE_TPL_BLAS``` for ```trtri``` but ```KOKKOSKERNELS_ENABLE_TPL_LAPACK``` still exists in the Makefile.kokkos-kernels. Are we still need ```KOKKOSKERNELS_ENABLE_TPL_LAPACK``` somewhere else?

Another comment is have you run spot-check with MAGMA enabled?

e10harvey commented:
> Another comment is have you run spot-check with MAGMA enabled?

No, I followed the guidelines in https://github.com/kokkos/kokkos-kernels/wiki/TestAllSandia#full-testing. I'm trying spot-check with magma now on kokkos-kernels:develop branch. If that passes, I will try spot-check with magma on e10harvey:issue-589.

> Are we still need `KOKKOSKERNELS_ENABLE_TPL_LAPACK` somewhere else?

It's not needed. I left it in because the LAPACKE target in Makefile.kokkos-kernels was causing link errors for me due to adding `-llapack`. @vqd8a: given this, let me know if you still want `KOKKOSKERNELS_ENABLE_TPL_LAPACK` removed from Makefile.kokkos-kernels.

e10harvey commented:
> I'm trying spot-check with magma now on kokkos-kernels:develop branch.

@vqd8a: I just realized this won't pass spot-check on any branch because magma depends on cuda and all spot-checks attempt non-cuda builds. Maybe this is why `--with-tpls` is missing from the `test_all_sandia` help menu. I did verify unit tests with blas,cublas,magma -- see attached files.
[cuda.all.LayoutLeft,LayoutRight.txt](https://github.com/kokkos/kokkos-kernels/files/4300373/cuda.all.LayoutLeft.LayoutRight.txt)
[openmp.all.LayoutLeft,LayoutRight.txt](https://github.com/kokkos/kokkos-kernels/files/4300374/openmp.all.LayoutLeft.LayoutRight.txt)

ndellingwood commented:
@e10harvey testing should now use the cm_test_all_sandia script with VOTD kokkos develop branch, test_all_sandia is no longer being maintained as it relies on the Makefile build system rather than CMake. The documentation will need to be updated.

e10harvey commented:
> @e10harvey testing should now use the cm_test_all_sandia script with VOTD kokkos develop branch, test_all_sandia is no longer being maintained as it relies on the Makefile build system rather than CMake. The documentation will need to be updated.

@ndellingwood: Thanks for the heads up. I'll re-run spot-check with the new script. I have been using kokkos 2.9.00 and building everything with the old Makefile build system. Do we just use the current HEAD of kokkos:develop for testing -- I don't see a `VOTD` tag. Also, can one pass spot-check with the magma, blas, and cublas tpls enabled on white using `cm_test_all_sandia`?

vqd8a commented:
@e10harvey I think you can run spot-check with MAGMA enabled on  ```apollo```. I installed ```magma/2.5.0/gcc/6.1.0/openblas/0.3.4/cuda/10.0``` on ```apollo``` when I tested ```gesv```.

e10harvey commented:
> @e10harvey I think you can run spot-check with MAGMA enabled on `apollo`. I installed `magma/2.5.0/gcc/6.1.0/openblas/0.3.4/cuda/10.0` on `apollo` when I tested `gesv`.

@vqd8a: My mistake. You're just looking for spot-check + magma with a single toolchain -- running now.

ndellingwood commented:
@e10harvey 

> Do we just use the current HEAD of kokkos:develop for testing

Correct


> Also, can one pass spot-check with the magma, blas, and cublas tpls enabled on white using `cm_test_all_sandia`

magma isn't available on white, but I see @vqd8a responded to point out that apollo has the tpl available.

e10harvey commented:
@vqd8a, @srajama1:
Here is the spot-check with magma enabled:
```bash
Going to test compilers:  cuda/10.0
Testing compiler cuda/10.0
  Starting job cuda-10.0-Cuda_OpenMP-release
  PASSED cuda-10.0-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
#######################################################
cuda-10.0-Cuda_OpenMP-release build_time=1088 run_time=1099
#######################################################
FAILED TESTS
#######################################################
```

As well as a spot-check with the new cmake build system:
```bash
Repository Status:  b9cb21185449b2bc7ed91c84aa322b51dac05290 blas/trtri: Fix macros names.


Going to test compilers:  gcc/6.4.0 gcc/7.2.0 ibm/16.1.0 cuda/9.2.88 cuda/10.1.105
Testing compiler gcc/6.4.0
Testing compiler gcc/7.2.0
  Starting job gcc-6.4.0-OpenMP_Serial-release
kokkos options: 
kokkos devices: OpenMP,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-6.4.0-OpenMP_Serial-release
  Starting job gcc-7.2.0-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
kokkos options: 
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.2.0-Serial-release
Testing compiler ibm/16.1.0
  Starting job gcc-7.2.0-OpenMP_Serial-release
kokkos options: 
kokkos devices: OpenMP,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.2.0-OpenMP_Serial-release
Testing compiler cuda/9.2.88
  Starting job ibm-16.1.0-Serial-release
kokkos options: 
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED ibm-16.1.0-Serial-release
  Starting job cuda-9.2.88-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-9.2.88-Cuda_OpenMP-release
Testing compiler cuda/10.1.105
  Starting job cuda-9.2.88-Cuda_Serial-release
kokkos options: 
kokkos devices: Cuda,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-9.2.88-Cuda_Serial-release
  Starting job cuda-10.1.105-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-10.1.105-Cuda_OpenMP-release
  Starting job cuda-10.1.105-Cuda_Serial-release
kokkos options: 
kokkos devices: Cuda,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-10.1.105-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.1.105-Cuda_OpenMP-release build_time=336 run_time=135
cuda-10.1.105-Cuda_Serial-release build_time=350 run_time=186
cuda-9.2.88-Cuda_OpenMP-release build_time=313 run_time=172
cuda-9.2.88-Cuda_Serial-release build_time=322 run_time=221
gcc-6.4.0-OpenMP_Serial-release build_time=115 run_time=158
gcc-7.2.0-OpenMP-release build_time=78 run_time=53
gcc-7.2.0-OpenMP_Serial-release build_time=124 run_time=155
gcc-7.2.0-Serial-release build_time=75 run_time=98
ibm-16.1.0-Serial-release build_time=385 run_time=110
```

vqd8a commented:
Thank you @e10harvey

srajama1 commented:
Can I merge ?

srajama1 commented:
BTW, why is this marked 3.2 and not 3.1 ?

e10harvey commented:
> Can I merge ?

Yes. Let me know if you want me to squash the commits down before merging.

> BTW, why is this marked 3.2 and not 3.1 ?

@ndellingwood and I discussed that the fall-back implementation of `trmm` and `trtri` would not be ready for 3.1.

srajama1 commented:
I see what you mean. I think this commit should still go in 3.1. The fallback mechanism can be 3.2.

ndellingwood commented:
This is causing build failures with clang/7+cuda/9 and gcc/5.3, I'll open up an issue. If it may take awhile to fix we can consider reverting the PR. @e10harvey
